pkg/rpcservice/listen.go

```diff
@@ -0,0 +1,200 @@
+		ForwardHeaders:        req.Headers,
+		ForwardConnectURL:     req.ForwardConnectTo,
+		ForwardConnectHeaders: req.ConnectHeaders,
+		UseConfiguredWebhooks: req.UseConfiguredWebhooks,
+		WebSocketFeature:      webhooksWebSocketFeature,
+		UseLatestAPIVersion:   req.Latest,
+		SkipVerify:            req.SkipVerify,
+		Log:                   logger,
+		Events:                req.Events,
+		OutCh:                 proxyOutCh,
+
+		// Hidden for debugging
+		APIBaseURL: "",
+		NoWSS:      false,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go p.Run(ctx)
+
+	for {
+		select {
+		case e := <-proxyOutCh:
+			err := e.Accept(proxyVisitor)
+			if err != nil {
+				return err
+			}
+		case <-stream.Context().Done():
+			return stream.Context().Err()
+		}
+	}
+}
+
+func createProxyVisitor(stream *rpc.StripeCLI_ListenServer) *websocket.Visitor {
+	return &websocket.Visitor{
+		VisitError: func(ee websocket.ErrorElement) error {
+			switch ee.Error.(type) {
+			case proxy.FailedToPostError, proxy.FailedToReadResponseError:
+				// These errors shouldn't end the stream
+				(*stream).Send(&rpc.ListenResponse{
+					Content: &rpc.ListenResponse_EndpointResponse_{
+						EndpointResponse: &rpc.ListenResponse_EndpointResponse{
+							Content: &rpc.ListenResponse_EndpointResponse_Error{
+								Error: ee.Error.Error(),
+							},
+						},
+					},
+				})
+				return nil
+			default:
+				return ee.Error
+			}
+		},
+		VisitData: func(de websocket.DataElement) error {
+			switch data := de.Data.(type) {
+			case proxy.StripeEvent:
+				eventData, err := structpb.NewStruct(data.Data)
+				if err != nil {
+					return err
+				}
+				request := rpc.StripeEvent_Request{
+					Id:             data.Request.ID,
+					IdempotencyKey: data.Request.IdempotencyKey,
+				}
+				(*stream).Send(&rpc.ListenResponse{
+					Content: &rpc.ListenResponse_StripeEvent{
+						StripeEvent: &rpc.StripeEvent{
+							Account:         data.Account,
+							ApiVersion:      data.APIVersion,
+							Created:         int64(data.Created),
+							Data:            eventData,
+							Id:              data.ID,
+							Type:            data.Type,
+							Livemode:        data.Livemode,
+							PendingWebhooks: int64(data.PendingWebhooks),
+							Request:         &request,
+						},
+					},
+				})
+				return nil
+			case proxy.EndpointResponse:
+				var httpMethodResponse rpc.ListenResponse_EndpointResponse_Data_HttpMethod
+				switch data.Resp.Request.Method {
+				case http.MethodDelete:
+					httpMethodResponse = rpc.ListenResponse_EndpointResponse_Data_HTTP_METHOD_DELETE
+				case http.MethodGet:
+					httpMethodResponse = rpc.ListenResponse_EndpointResponse_Data_HTTP_METHOD_GET
+				case http.MethodPost:
+					httpMethodResponse = rpc.ListenResponse_EndpointResponse_Data_HTTP_METHOD_POST
+				default:
+					httpMethodResponse = rpc.ListenResponse_EndpointResponse_Data_HTTP_METHOD_UNSPECIFIED
+				}
```

Comment:
nit but I wonder if this could be a bit cleaner with a map and getter function, so we could load the correct rpc method inline

```
HttpMethod: getRPCMethodFromRequestMethod(data.Resp.Request.Method)
```

Or even further, could we have a `buildListenResponse` method that takes data and build the whole thing, so the case statement is a bit cleaner?

```
VisitData: func(de websocket.DataElement) error {
    switch data := de.Data.(type) {
    case proxy.StripeEvent:
        (*stream).Send(buildStripeEventResponse(data))
    case proxy.EndpointResponse:
        (*stream).Send(buildEndpointResponse(data))
    default:
        [...]
```

Reply:
would also make testing a bit simpler as each independent function can be tested in isolation